stratum/src/lib.rs

```diff
@@ -161,31 +240,50 @@ impl Stratum {
 
 	pub fn maintain(&self) {
 		let mut job_que = self.job_que.write();
-		let workers = self.workers.read();
+		let job_payload = self.dispatcher.job();
 		for socket_addr in job_que.drain() {
-			if let Some(worker_id) = workers.get(&socket_addr) {
-				let job_payload = self.dispatcher.job(worker_id.to_owned());
-				job_payload.map(
-					|json| self.rpc_server.push_message(&socket_addr, json.as_bytes())
-				);
-			}
-			else {
-				trace!(
-					target: "stratum",
-					"Job queued for worker that is still not authorized, skipping ('{:?}')", socket_addr
-				);
-			}
+			job_payload.as_ref().map(
+				|json| self.tcp_dispatcher.push_message(&socket_addr, json.to_owned())
+			);
 		}
 	}
 }
 
 impl PushWorkHandler for Stratum {
 	fn push_work_all(&self, payload: String) -> Result<(), Error> {
-		let workers = self.workers.read();
-		println!("pushing work for {} workers", workers.len());
-		for (ref addr, _) in workers.iter() {
-			self.rpc_server.push_message(addr, payload.as_bytes())?;
+		let hup_peers = {
+			let workers = self.workers.read();
+			let next_request_id = {
+				let mut counter = self.notify_counter.write();
+				if *counter == ::std::u32::MAX { *counter = NOTIFY_CONTER_INITIAL; }
+				else { *counter = *counter + 1 }
+				*counter
+			};
+
+			let mut hup_peers = HashSet::with_capacity(0); // most of the cases won't be needed, hence avoid allocation
```

Comment:
this behavior can change, i explicitly state that i don't need any space for elements with high probability